File: Maximum_Gcd_and_Sum.java
import java.util.*;

public class Maximum_Gcd_and_Sum {
    
    // Function to get all factors of a number
    private static List<Integer> getFactors(int num) {
        List<Integer> factors = new ArrayList<>();
        for (int i = 1; i * i <= num; i++) {
            if (num % i == 0) {
                factors.add(i);
                if (i != num / i) factors.add(num / i);
            }
        }
        return factors;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int[] A = new int[n];
        int[] B = new int[n];

        for (int i = 0; i < n; i++) A[i] = sc.nextInt();
        for (int i = 0; i < n; i++) B[i] = sc.nextInt();

        // Step 1: Build factor map for A
        Map<Integer, Integer> factorMap = new HashMap<>(); 
        // (factor -> max number in A having this factor)
        
        for (int x : A) {
            for (int f : getFactors(x)) {
                factorMap.put(f, Math.max(factorMap.getOrDefault(f, 0), x));
            }
        }

        int maxGcd = 0;
        int maxSum = 0;

        // Step 2: For each element in B, check its factors
        for (int y : B) {
            for (int f : getFactors(y)) {
                if (factorMap.containsKey(f)) {
                    int x = factorMap.get(f);
                    int sum = x + y;
                    if (f > maxGcd || (f == maxGcd && sum > maxSum)) {
                        maxGcd = f;
                        maxSum = sum;
                    }
                }
            }
        }

        System.out.println(maxSum);
        sc.close();
    }
}

File: Maximum_Gcd_and_Sum.cpp
#include <bits/stdc++.h>
using namespace std;

// Function to get all factors of a number
vector<int> getFactors(int n) {
    vector<int> factors;
    for (int i = 1; i * i <= n; i++) {
        if (n % i == 0) {
            factors.push_back(i);
            if (i != n / i)
                factors.push_back(n / i);
        }
    }
    return factors;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    cin >> n;
    vector<int> A(n), B(n);

    for (int i = 0; i < n; i++) cin >> A[i];
    for (int i = 0; i < n; i++) cin >> B[i];

    // Step 1: Build factor map for A
    unordered_map<int, int> factorMap;  // factor -> max element in A with that factor
    for (int x : A) {
        vector<int> factors = getFactors(x);
        for (int f : factors) {
            auto it = factorMap.find(f);
            if (it == factorMap.end() || it->second < x)
                factorMap[f] = x;
        }
    }

    int maxGCD = 0, maxSum = 0;

    // Step 2: Check for each element in B
    for (int y : B) {
        vector<int> factors = getFactors(y);
        for (int f : factors) {
            auto it = factorMap.find(f);
            if (it != factorMap.end()) {
                int x = it->second;
                int sum = x + y;
                if (f > maxGCD || (f == maxGCD && sum > maxSum)) {
                    maxGCD = f;
                    maxSum = sum;
                }
            }
        }
    }

    cout << maxSum << "\n";
    return 0;
}
